New draft: The function that changed my mind

sierisimo opened the pull request:
Adding a new article based on my –already presented a lot– talk about a refactor to a single block of code.

Also thinking on splitting into 2 articles. 

The file is README-en.md because I'm thinking on translate to spanish or create an spanish version of this.

Resolves #9

sierisimo committed:
New draft: The function that changed my mind

sierisimo committed:
Prettier changer for markdown formatting

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+---
+title: The Function that changed my mind
+published: false
+description: How a single function show me a lot of kotlin
```
```suggestion
description: How a single function showed me a lot of Kotlin
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+series:
+---
+
+When writting code we should always have a place for improvement. Improvement can present itself in different ways or forms. A proof of it is how in recent years the big attention to CI (_Continous Integration_) and CD (_Continous Delivery_) has grown. Same goes for new languages and technologies.
```
```suggestion
When writting code, we should always have a place for improvement. Improvement can present itself in different ways. A proof of it is how in recent years the attention to CI (_Continuous Integration_) and CD (_Continuous Delivery_) has grown significantly. Same goes for new languages and technologies.
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+**Disclaimer**: If you know me in person or have been in a meetup with me presenting, there's a big chance you already heard me talking about this, even [on YouTube with the GDG Santo Domingo]() or Kotlin/Everywhere Guadalajara and Mexico City.
+
+**Another disclaimer**: This article is not about architecture –even when it's a hot topic in the Android community– or best practices or even about how to organize an Android Aplication, it's just about a set of changes that showed me some concepts of Kotlin. Also, I started this changes almost 3 years ago, by the point of publication of this article, probably the libraries from Google –like `co ...
```
```suggestion
**Another disclaimer**: This article is not about architecture –even when it's a hot topic in the Android community– or best practices or even about how to organize an Android Application. It's just about a set of changes that showed me some Kotlin concepts. Also, I started these changes almost 3 years ago, so by the time this article is published, the libraries from Google –like `core-ktx`– will probably already have something similar or better. And just to be clear: None of the approaches discussed here are right or wrong, all the solutions depend on individual needs and individual reasons to choose them over the others.
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+## But first… some Context
+
+If you're an Android Developer or have worked with Android, probably you catch the pun intended in the joke, if not, let me explain you quickly.
```
```suggestion
If you're an Android Developer or have worked with Android, you probably caught the pun intended by this title. If not, let me explain it to you quickly.
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+**Another disclaimer**: This article is not about architecture –even when it's a hot topic in the Android community– or best practices or even about how to organize an Android Aplication, it's just about a set of changes that showed me some concepts of Kotlin. Also, I started this changes almost 3 years ago, by the point of publication of this article, probably the libraries from Google –like `co ...
+
+## But first… some Context
```
```suggestion
## But first… some _Context_
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+In Android, everything happening on your application has a `Context` and to make some stuff work, you also need a `Context`. This is just a class that Android uses to give information and access to the app.
+
+One specific case is when you want to start a new `Activity` on your app. This means –in short– launching a new whole screen or launching an external application. To do this, you first create an `Intent` based on your `Context` and pass the `Actiity` class you want to start:
```
```suggestion
One specific case is when you want to start a new `Activity` on your app. This means –in short– launching a whole new screen or an external application. To do this, you first create an `Intent` based on your `Context` and pass the `Activity` class you want to start:
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+Then you can start `MyNextActivity` from another `Activity` simply using this intent: `startActivity(myIntent)`
+
+If your application relays on activities for showing different sections/views of your application then probably you will end with a lot of this calls distributed around. In some different variants, but all of them with the same structure:
```
```suggestion
If your application relays on activities for showing different sections/views of your application then probably you will end with a lot of these calls distributed around. In some different variants, but all of them with the same structure:
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
````diff
@@ -0,0 +1,322 @@
+}
+```
+
+You can add different layers of complexity but this is the base for most solutions: **have a single point for launching/starting new activities**. In my case I always ended having case 1 all around. Having functions here and there seemed like a good solution, until I ended with 120 functions doing almost the same all around, which is not bad but seeing that `::class.java` even inside of different ...
````
```suggestion
You can add different layers of complexity but this is the base for most solutions: **have a single point for launching/starting new activities**. In my case, I always ended having case 1 all around. Having functions here and there seemed like a good solution, until I ended with 120 functions doing almost the same all around, which is not bad, but seeing that `::class.java` even inside of different methods really bothered me. Plus, seeing 3 or 4 functions inside the same class with almost the same body also feels wrong.
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+But wait… _most_? We want **all**, not _most_… also this has other issues:
+
+1. Is not always possible to inherit
```
```suggestion
1. It is not always possible to inherit
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+1. Is not always possible to inherit
+2. Not all the classes in need of the code will be a kind of `Activity`
+3. When some class has access to `Activity` it will need to cast in an ugly way
```
```suggestion
3. When some class has access to `Activity` it will need to be cast in an ugly way
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+1. Is not always possible to inherit
+2. Not all the classes in need of the code will be a kind of `Activity`
+3. When some class has access to `Activity` it will need to cast in an ugly way
+4. The `::class.java` is still present –ugh, it's ugly–
```
```suggestion
4. `::class.java` is still present –ugh, it's ugly!
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
````diff
@@ -0,0 +1,322 @@
+}
+```
+
+This is not bad but is a caveat of having Kotlin extension functions and… well, if is only to make things readable, we are not making too much progress, we only removed 1 line of code from our original block of code. What can save us from actually having this on the bytecode and still adding readability to our codebase?
````
```suggestion
This is not bad but is a caveat of having Kotlin extension functions and… well, if it is only to make things readable, we are not making too much progress: we only removed 1 line of code from our original block of code. What can save us from actually having this on the bytecode and still adding readability to our codebase?
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+## Lambda with reciever
+
+Because we don't want to increment the size of our parameter list, because the types of parameters can grow and because we don't want to have multiple wrapper functions just for the parameters, we need to find a way to "_configure_" or `Intent` _on demand_.
```
```suggestion
Because we don't want to increment the size of our parameter list, because the types of parameters can grow and because we don't want to have multiple wrapper functions just for the parameters, we need to find a way to "_configure_" our `Intent` _on demand_.
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
````diff
@@ -0,0 +1,322 @@
+}
+```
+
+Adding this last parameter we are saying that our function can take a lambda that will execute in the scope of an `Intent` and can use properties and methods from the `Intent`. This lambda is later used to setup our `Intent` and finally we start the activity with the already configured `Intent`.
````
```suggestion
Adding this last parameter we are saying that our function can take a lambda that will execute in the scope of an `Intent` and can use properties and methods from the `Intent` object. This lambda is later used to set up our `Intent` and finally we start the activity with the already configured `Intent`.
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+As you can see, no reference to external classes, no calls to external functions, all living there. This is a good thing, as the code stays simple while coding and has the real meaning when executing.
+
+But still… while codign the `::class.java` is present. Why I have been complaining about this if is something present in Kotlin? Well, the main reason is because this operator and class is not nice, it's quite ugly and also can be confusing. There are more than one way to obtain a `Class<T>` from a class and this one is the easier one. There's a chance for erro using the other ones and they are v ...
```
```suggestion
But still… while coding, the `::class.java` is present. Why have I been complaining about this if it is something present in Kotlin? Well, the main reason is because this operator and class is not nice, it's quite ugly and also can be confusing. There is more than one way to obtain a `Class<T>` from a class and this one is the easier one. There's a chance for error using the other ones and they are verbose, not what we are trying to avoid here.
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+Some final touches to the signature will make the function even safer and even cooler when being called:
+
+1. Make `T` have a constraint to `Activity` in a way only classes that extend from `Activity` are allowed
+2. Add a default value to the _lambda with reciever_ to allow calls with empty configuration
```
```suggestion
2. Add a default value to the _lambda with receiver_ to allow calls with an empty configuration
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
````diff
@@ -0,0 +1,322 @@
+}
+```
+
+Now our function is cooler than ever, won't add too much to the final bytecode and also will make our code more readable. Wow! We traveled a long way doing a lot of changes to a single piece of code to make things cooler, but not just that, now other people developing along with us can benefit from this improvement as they can simply call our function without having to worry how it works. I have  ...
````
```suggestion
Now our function is cooler than ever, won't add too much to the final bytecode and also will make our code more readable. Wow! We traveled a long way doing a lot of changes to a single piece of code to make things cooler, but not just that, now other people developing along with us can benefit from this improvement as they can simply call our function without having to worry how it works. I have seen even people adding animations, making a way to hold results and transforming data using these kind of functions/approaches. Writing idiomatic Kotlin actually makes you think and develop different and cooler!
```

jhoon commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+## Conlusion
+
+We can apply this kind of approch by steps or go directly to other functions on our codebase, one example is the usage of `SharedPreferences` and making the _edition_ do an _auto-apply_ "magically" using the lambda with reciever. We can create ways to configure more stuff in less steps and this will make our writting of code improving as we are adding tools to our set of _already-vaste_ set of fu ...
```
```suggestion
We can apply this kind of approch by steps or go directly to other functions on our codebase, one example is the usage of `SharedPreferences` and making the _edition_ do an _auto-apply_ "magically" using the lambda with receiver. We can create ways to configure more stuff in less steps and this will make our writing of code improve as we will be adding tools to our _already-vast_ set of functions and tools on Kotlin. Thinking about what is repeated or can be improved allows us –the developers– to discover new worlds and make things cooler and sharper. Don't be scared of change.
```

jhoon reviewed (changes requested):
A bunch of typos, but also some wording adjustments :)

sierisimo committed:
Update TheFunctionChangedMyMind/README-en.md

Co-Authored-By: Jhoon Saravia <jhoon@users.noreply.github.com>

sierisimo committed:
Apply suggestions from code review

Co-Authored-By: Jhoon Saravia <jhoon@users.noreply.github.com>

Commit:
Adding information for the TravisCI

Commit:
Adding link to youtube video

also adding a promise for a next article on reified
also changing to true the published status for the moment it goes merged

Commit:
Adding disclaimer of language

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+title: The Function that changed my mind
+published: true
+description: How a single function showed me a lot of Kotlin
+tags: kotlin,functions,android,refactor
```
Add space after each comma

sierisimo replied:
I won't. It's not necessary for the tag system and in some poor designed tags systems it would cause errors. this is just metadata

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+series:
+---
+
+When writting code, we should always have a place for improvement. Improvement can present itself in different ways. A proof of it is how in recent years the attention to CI (_Continuous Integration_) and CD (_Continuous Delivery_) has grown significantly. Same goes for new languages and technologies.
```
A proof about it, is the way the attention of CI and CD has significantly grown in recent years.

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+When writting code, we should always have a place for improvement. Improvement can present itself in different ways. A proof of it is how in recent years the attention to CI (_Continuous Integration_) and CD (_Continuous Delivery_) has grown significantly. Same goes for new languages and technologies.
+
+But not all involve changing to a new language, technology or tool. Some of these improvements are simply changing how something works, adding tests, adding abstractions, changing patterns, etc. This is the case for the change I introduced on one of my projects.
```
Nevertheless, this doesn't involve changing to a new language, technology, nor tool. Some of these improvements are so simple as changing how something works, i. e. adding tests/abstractions, changing patterns, etc. This is how I have proceeded in several projects I have worked in.

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+But not all involve changing to a new language, technology or tool. Some of these improvements are simply changing how something works, adding tests, adding abstractions, changing patterns, etc. This is the case for the change I introduced on one of my projects.
+
+**Disclaimer**: If you know me in person or have been in a meetup with me presenting, there's a big chance you already heard me talking about this, even [on YouTube with the GDG Santo Domingo (in spanish)](https://www.youtube.com/watch?v=0KoXxoPHV4o&feature=youtu.be) or Kotlin/Everywhere Guadalajara and Mexico City.
```
Whether you have talked with me in person, or have attended to one of my talks at meetups, it's likely, you already have heard me talking about it, even....

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+**Disclaimer**: If you know me in person or have been in a meetup with me presenting, there's a big chance you already heard me talking about this, even [on YouTube with the GDG Santo Domingo (in spanish)](https://www.youtube.com/watch?v=0KoXxoPHV4o&feature=youtu.be) or Kotlin/Everywhere Guadalajara and Mexico City.
+
+**Another disclaimer**: This article is not about architecture –even when it's a hot topic in the Android community– or best practices or even about how to organize an Android Application. It's just about a set of changes that showed me some Kotlin concepts. Also, I started these changes almost 3 years ago, so by the time this article is published, the libraries from Google –like `core-ktx`– will ...
```
–even though it's a hot topic in the Android community –, best practices or even about how...

Also, I started to implement this aforementioned strategies almost 3 years ago, therefor by the time.... 

will probably being updated or improved.

And just to be clear: It's good to mention that the approaches mentioned in the following lines don't apply to all scenarios nor cases; the solutions mentioned in this article depend on scenarios or needs themselves, besides individual reasons...

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+If you're an Android Developer or have worked with Android, you probably caught the pun intended by this title. If not, let me explain it to you quickly.
+
+In Android, everything happening on your application has a `Context` and to make some stuff work, you also need a `Context`. This is just a class that Android uses to give information and access to the app.
```
In Android development, everything happening on your application has a `Context` and to make some stuff work, you need to deal with a `Context` as well. This is no more than a class that Android uses to provide information and access to the app.

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+In Android, everything happening on your application has a `Context` and to make some stuff work, you also need a `Context`. This is just a class that Android uses to give information and access to the app.
+
+One specific case is when you want to start a new `Activity` on your app. This means –in short– launching a whole new screen or an external application. To do this, you first create an `Intent` based on your `Context` and pass the `Activity` class you want to start:
```
One specific case arises at the time it's required to start a new `Activity` on your app. This means -succinctly- ...

In order to accomplish it, firstly you should create an `Intent` ... pass the `Activity` you would like to launch:

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+Then you can start `MyNextActivity` from another `Activity` simply using this intent: `startActivity(myIntent)`
+
+If your application relays on activities for showing different sections/views of your application then probably you will end with a lot of these calls distributed around. In some different variants, but all of them with the same structure:
```
//Relays is not relies, keep that in mind

If your application relies on activities for showing varied sections or views of your application, then probably you will end **up** with a lot of these calls distributed around. All of them with the same structure yet with several variants, such as:

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+## My first problem
+
+To be honest, there is no problem with this code. Not at all. But I have seen this happening in a lot of applications, you start having these calls distributed around without control, then you suddenly have 2 or 3 places doing the same, setting the same intent with a different variable name but same parameters and then doing the same call. To solve these, most people usually do one of two things:
```
...a lot of applications, firstly you have these calls distributed around without messed around, then you suddenly have 2 or 3 repetitions of the same work in different places...

...the same call. To overcome these...

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+
+To be honest, there is no problem with this code. Not at all. But I have seen this happening in a lot of applications, you start having these calls distributed around without control, then you suddenly have 2 or 3 places doing the same, setting the same intent with a different variable name but same parameters and then doing the same call. To solve these, most people usually do one of two things:
+
+1. Wrap it around a single method for each case
```
Wrap it around a single method for each case **.**

sierisimo replied:
I know you hate it, but consistency is the real MVP. No dot at the end of list items.

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
```diff
@@ -0,0 +1,322 @@
+To be honest, there is no problem with this code. Not at all. But I have seen this happening in a lot of applications, you start having these calls distributed around without control, then you suddenly have 2 or 3 places doing the same, setting the same intent with a different variable name but same parameters and then doing the same call. To solve these, most people usually do one of two things:
+
+1. Wrap it around a single method for each case
+2. Put all under a single class/object and just pass needed parameters for repeated cases
```
2. Put all under a single class/object and just pass needed parameters for repeated cases **.**

AnnaElizleal commented on TheFunctionChangedMyMind/README-en.md:
````diff
@@ -0,0 +1,322 @@
+}
+```
+
+You can add different layers of complexity but this is the base for most solutions: **have a single point for launching/starting new activities**. In my case, I always ended having case 1 all around. Having functions here and there seemed like a good solution, until I ended with 120 functions doing almost the same all around, which is not bad, but seeing that `::class.java` even inside of differe ...
````
You can add several layers of complexity, but this is the base for most solutions:

In my case, I always ended **up** with 120...

sierisimo commented:
That's not actually relevant for tags. It can be even dangerous if the tag
system is poor designed

On Mon, Sep 16, 2019, 1:02 PM Anna Eliz <notifications@github.com> wrote:

> *@AnnaElizleal* commented on this pull request.
> ------------------------------
>
> In TheFunctionChangedMyMind/README-en.md
> <https://github.com/sierisimo/publications/pull/8#discussion_r324809968>:
>
> > @@ -0,0 +1,322 @@
> +---
> +title: The Function that changed my mind
> +published: true
> +description: How a single function showed me a lot of Kotlin
> +tags: kotlin,functions,android,refactor
>
> Add space after each comma
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sierisimo/publications/pull/8?email_source=notifications&email_token=AAM7AJOIZOGCSKJ5ZZEM2ODQJ7C2BA5CNFSM4IUX3U62YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCE3MOUY#pullrequestreview-288802643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAM7AJOIVTXJMHDKMFC6VCDQJ7C2BANCNFSM4IUX3U6Q>
> .
>

sierisimo committed:
Fixes from @annaelizleal

sierisimo committed:
Touches suggested by prettier

AnnaElizleal reviewed (approved):
Comments added
